File: casper/src/test/scala/coop/rchain/casper/helper/Secp256k1SignContract.scala
package coop.rchain.casper.helper

import cats.effect.Concurrent
import coop.rchain.crypto.signatures.Secp256k1
import cats.syntax.functor._
import coop.rchain.metrics.Span
import coop.rchain.models.{ListParWithRandom, Par}
import coop.rchain.rholang.interpreter.SystemProcesses
import coop.rchain.rholang.interpreter.{ContractCall, RhoType}

object Secp256k1SignContract {

  def get[F[_]: Concurrent: Span](
      ctx: SystemProcesses.ProcessContext[F]
  )(message: Seq[ListParWithRandom], isReplay: Boolean, previousOutput: Seq[Par]): F[Seq[Par]] = {
    val isContractCall = new ContractCall(ctx.space, ctx.dispatcher)
    (message, isReplay, previousOutput) match {
      case isContractCall(
          produce,
          _,
          _,
          Seq(RhoType.ByteArray(hash), RhoType.ByteArray(sk), ackCh)
          ) =>
        val sig    = Secp256k1.sign(hash, sk)
        val output = Seq(RhoType.ByteArray(sig))
        produce(output, ackCh)
    }
  }
}

File: casper/tests/helper/secp256k1_sign_contract.rs
// See casper/src/test/scala/coop/rchain/casper/helper/Secp256k1SignContract.scala

use k256::ecdsa::{signature::hazmat::PrehashSigner, Signature, SigningKey};
use models::{
    rhoapi::{ListParWithRandom, Par},
    rust::utils::new_gbytearray_par,
};
use rholang::rust::interpreter::{
    errors::illegal_argument_error, errors::InterpreterError, rho_type::RhoByteArray,
    system_processes::ProcessContext,
};

use crate::helper::process_context_ext::ProcessContextExt;

pub async fn get(
    ctx: ProcessContext,
    message: (Vec<ListParWithRandom>, bool, Vec<Par>),
) -> Result<Vec<Par>, InterpreterError> {
    let is_contract_call = ctx.contract_call();

    if let Some((produce, _, _, args)) = is_contract_call.unapply(message) {
        match args.as_slice() {
            [hash_par, sk_par, ack_channel] => {
                if let (Some(hash), Some(secret_key)) = (
                    RhoByteArray::unapply(hash_par),
                    RhoByteArray::unapply(sk_par),
                ) {
                    if secret_key.len() != 32 {
                        return Err(InterpreterError::BugFoundError(format!(
                            "Invalid private key length: must be 32 bytes, got {}",
                            secret_key.len()
                        )));
                    }

                    let signing_key =
                        SigningKey::from_slice(&secret_key).expect("Invalid private key");

                    let signature: Signature = signing_key
                        .sign_prehash(&hash)
                        .expect("Failed to sign prehash");
                    let der_bytes = signature.to_der().as_bytes().to_vec();

                    let result_par = new_gbytearray_par(der_bytes, Vec::new(), false);

                    let output = vec![result_par];
                    produce(&output, &ack_channel).await?;
                    Ok(output)
                } else {
                    Err(illegal_argument_error("secp256k1_sign"))
                }
            }
            _ => Err(illegal_argument_error("secp256k1_sign")),
        }
    } else {
        Err(illegal_argument_error("secp256k1_sign"))
    }
}
